pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,353 @@
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter("block_oids", fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter("block_oids", fmt.Sprintf("%d", startIndex), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			blockOidsObject, err := bfs.search(prm)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to find 'block_oids' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+			if len(blockOidsObject) == 0 {
+				bfs.log.Info(fmt.Sprintf("No 'block_oids' object found with index %d, stopping.", startIndex))
+				return errors.New("no 'block_oids' object found")
+			}
+
+			blockOidsData, err := bfs.get(blockOidsObject[0].String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to fetch 'block_oids' object %d", startIndex), zap.Error(err))
+				return err
+			}
+			blockOIDs, err := parseBlockOIDs(blockOidsData)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to parse 'block_oids' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			for j, oidBlock := range blockOIDs {
+				if j < int(skip) {
+					continue
+				}
+				data, err := bfs.get(oidBlock.String())
+				if err != nil {
+					bfs.log.Error(fmt.Sprintf("Failed to fetch block with OID %s", oidBlock.String()), zap.Error(err))
+					return err
+				}
+				bfs.log.Info(fmt.Sprintf("Fetched block with OID %s", oidBlock.String()))
+				err = bfs.processBlock(data)
+				if err != nil {
+					bfs.log.Error(fmt.Sprintf("Failed to process blocks with index %d", startIndex), zap.Error(err))
+					return err
+				}
+			}
+
+			startIndex++
+		}
+	}
+}
+
+func (bfs *Service) processBlock(data []byte) error {
+	var buf []byte
+	stateRootInHeader := bfs.chain.GetConfig().StateRootInHeader
+	br := gio.NewBinReaderFromBuf(data)
+
+	readBlock := func(r *gio.BinReader) ([]byte, error) {
+		var size = r.ReadU32LE()
+		if uint32(cap(buf)) < size {
+			buf = make([]byte, size)
+		} else {
+			buf = buf[:size]
+		}
+		r.ReadBytes(buf)
+		return buf, r.Err
+	}
+
+	buf, err := readBlock(br)
+	if err != nil {
+		return err
+	}
+	b := block.New(stateRootInHeader)
+	r := gio.NewBinReaderFromBuf(buf)
+	b.DecodeBinary(r)
+	err = bfs.queue.PutBlock(b)
+	if err != nil {
+		bfs.log.Error("failed to queue block", zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+func (bfs *Service) processHeaders(data []byte) error {
+	var resHeader payload.Headers
+	br := gio.NewBinReaderFromBuf(data)
+	resHeader.DecodeBinary(br)
+	err := bfs.chain.AddHeaders(resHeader.Hdrs...)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (bfs *Service) get(oid string) ([]byte, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+	defer cancel()
+	u, err := url.Parse(fmt.Sprintf("neofs:%s/%s", bfs.cfg.ContainerID, oid))
+	if err != nil {
+		return nil, err
+	}
+	rc, err := neofs.GetWithClient(ctx, bfs.client, bfs.privateKey, u, bfs.cfg.Addresses[0])
```

Comment:
Don't forget to close the ReadCloser after all data are fetched, it's important to release resources.